File: LMJPageControl/LMJPageControl.m
//
//  LMJPageControl.m
//  PageControl
//
//  Version 1.0.0
//
//  Created by MajorLi on 14-12-5.
//  Copyright (c) 2014年 iOS开发者公会. All rights reserved.
//
//  iOS开发者公会-技术1群 QQ群号：87440292
//  iOS开发者公会-技术2群 QQ群号：232702419
//  iOS开发者公会-议事区  QQ群号：413102158
//

#import "LMJPageControl.h"

@implementation LMJPageControl
{
    CGFloat _pointWidth;
    
    UIColor * _normalColor;
    UIColor * _selectedColor;
    
    UIImage * _normalImg;
    UIImage * _selectedImg;
    
    BOOL _isImagePoint;
    
    BOOL _isSetPagesNumber;
}

- (id)init{
    NSAssert(NO, @"禁止使用init函数进行控件初始化，请使用initWithFrame:pointSize:normalImage:selectedImage:或initWithFrame:pointSize:normalColor:selectedColor:函数进行控件初始化!");
    return nil;
}
- (id)initWithFrame:(CGRect)frame{
    NSAssert(NO, @"禁止使用initWithFrame:函数进行控件初始化，请使用initWithFrame:pointSize:normalImage:selectedImage:或initWithFrame:pointSize:normalColor:selectedColor:函数进行控件初始化!");
    return nil;
}
-(id)initWithFrame:(CGRect)frame pointSize:(CGFloat)width normalImage:(UIImage *)normalImg selectedImage:(UIImage *)selectedImg{
    self = [super initWithFrame:frame];
    if (self) {
        
        self.backgroundColor = [UIColor clearColor];
        _isSetPagesNumber = NO;
        
        _pointWidth = width;
        _normalImg = normalImg;
        _selectedImg = selectedImg;
        _isImagePoint = YES;
    }
    return self;
}

-(id)initWithFrame:(CGRect)frame pointSize:(CGFloat)width normalColor:(UIColor *)normalColor selectedColor:(UIColor *)selectedColor{
    self = [super initWithFrame:frame];
    if (self) {
        
        self.backgroundColor = [UIColor clearColor];
        _isSetPagesNumber = NO;
        
        _pointWidth = width;
        _normalColor = normalColor;
        _selectedColor = selectedColor;
        _isImagePoint = NO;
    }
    return self;
}

-(void)setNumberOfPages:(NSInteger)pages{
    _isSetPagesNumber = YES;
    
    for (int i = 0; i <self.subviews.count; i++) {
        [[self.subviews objectAtIndex:i] removeFromSuperview];
    }
    
    CGFloat distanceBetweenPoints = (self.frame.size.width -_pointWidth*pages)/(pages-1);
    
    if (_isImagePoint) {
        
        for (int i = 0; i< pages; i++) {
            UIImageView * imageView = [[UIImageView alloc] initWithFrame:CGRectMake((_pointWidth+distanceBetweenPoints)*i, (self.frame.size.height-_pointWidth)/2, _pointWidth, _pointWidth)];
            [imageView setImage:_normalImg];
            imageView.tag = 1000 + i;
            [self addSubview:imageView];
        }
    }else{
        for (int i = 0; i< pages; i++) {
            UIView * pointView = [[UIView alloc] initWithFrame:CGRectMake((_pointWidth+distanceBetweenPoints)*i, (self.frame.size.height-_pointWidth)/2, _pointWidth, _pointWidth)];
            [pointView setBackgroundColor:_normalColor];
            pointView.layer.cornerRadius = _pointWidth/2;
            pointView.tag = 2000 + i;
            [self addSubview:pointView];
        }
    }
}


-(void)setCurrentPage:(NSInteger)page{
    if (self.subviews.count == 0) {
        return;
    }
    
    if (_isImagePoint) {
        for (UIView * view in self.subviews) {
            
            UIImageView * imageView = (UIImageView *)view;
            
            if (imageView.tag - 1000 == page) {
                [imageView setImage:_selectedImg];
            }else{
                [imageView setImage:_normalImg];
            }
        }
    }else{
        for (UIView * view in self.subviews) {
            if (view.tag - 2000 == page) {
                [view setBackgroundColor:_selectedColor];
            }else{
                [view setBackgroundColor:_normalColor];
            }
        }
    }
}




/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
}
*/

@end

File: LMJPageControl/LMJPageControl.swift
//
//  LMJPageControl.swift
//  LMJPageControlExample
//
//  Created by LiMingjie on 2019/5/30.
//  Copyright © 2019 LMJ. All rights reserved.
//

import Foundation
import UIKit

open class LMJPageControl: UIView {
    fileprivate var _offsetLeft: Float!      = 0
    fileprivate var _offsetRight: Float!     = 0
    fileprivate var _pointWidth: Float!      = 5.0

    fileprivate var _normalColor: UIColor!   = UIColor.lightGray
    fileprivate var _selectedColor: UIColor! = UIColor.black
    fileprivate var _normalImg: UIImage?     = nil
    fileprivate var _selectedImg: UIImage?   = nil

    fileprivate var _pageCount: UInt!        = 1
    fileprivate var _currentPage: UInt!      = 0

    fileprivate var _points: Array           = [UIImageView]()

    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        reloadControl()
    }
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        reloadControl()
    }
    override open func layoutSubviews() {
        super.layoutSubviews()
        reloadPointsLayerout()
    }
    
    private func reloadControl() {
        reloadPoints()
        reloadPointsSize()
        reloadPointsStyle()
    }
    
    
    private func reloadPoints() {
        _points.forEach { point in
            point.removeFromSuperview()
        }
        _points.removeAll()
        for _ in 0..<_pageCount {
            let point = UIImageView()
            self.addSubview(point)
            _points.append(point)
        }
    }
    
    private func reloadPointsSize() {
        for i in 0..<_points.count {
            let point = _points[i]
            point.frame = CGRect(x: point.frame.origin.x, y: point.frame.origin.y, width: CGFloat(_pointWidth), height: CGFloat(_pointWidth))
            if i == _currentPage {
                if _selectedImg == nil {
                    point.layer.cornerRadius = CGFloat(_pointWidth / 2.0)
                } else {
                    point.layer.cornerRadius = 0.0
                }
            } else {
                if _normalImg == nil {
                    point.layer.cornerRadius = CGFloat(_pointWidth / 2.0)
                } else {
                    point.layer.cornerRadius = 0.0
                }
            }
            
        }
    }
    
    private func reloadPointsLayerout() {
        let fWidth = Float(self.frame.size.width)
        let fHeight = Float(self.frame.size.height)
        if fWidth <= 0 || fHeight <= 0 {
            return
        }
        
        let fPageCount = Float(_pageCount)
        let distance: Float = fWidth - _offsetLeft - _offsetRight - _pointWidth * fPageCount
        let distanceBetweenPoints: Float = fPageCount > 1.0 ? distance / (fPageCount-1.0) : 0.0;
        
        for i in 0..<_points.count {
            let point = _points[i]
            let fI = Float(i)
            let x = CGFloat(_offsetLeft + (_pointWidth + distanceBetweenPoints) * fI)
            let y = CGFloat((fHeight - _pointWidth) / 2.0)
            point.frame = CGRect(x: x, y: y, width: point.frame.size.width, height: point.frame.size.height)
        }
    }
    
    private func reloadPointsStyle() {
        for i in 0..<_points.count {
            let point = _points[i]
            if _currentPage == i {
                if _selectedImg != nil {
                    point.layer.masksToBounds = false
                    point.layer.cornerRadius  = 0.0
                    point.image               = _selectedImg!
                    point.backgroundColor     = _selectedColor
                } else {
                    point.layer.masksToBounds = true
                    point.layer.cornerRadius  = CGFloat(_pointWidth / 2.0)
                    point.image               = nil
                    point.backgroundColor     = _selectedColor
                }
            } else {
                if _normalImg != nil {
                    point.layer.masksToBounds = false
                    point.layer.cornerRadius  = 0.0
                    point.image               = _normalImg!
                    point.backgroundColor     = _normalColor
                } else {
                    point.layer.masksToBounds = true
                    point.layer.cornerRadius  = CGFloat(_pointWidth / 2.0)
                    point.image               = nil
                    point.backgroundColor     = _normalColor
                }
            }
        }
    }
    
    
    public var offsetLeft: Float {
        get {
            return _offsetLeft
        }
        set {
            _offsetLeft = newValue
            reloadPointsLayerout()
        }
    }
    public var offsetRight: Float {
        get {
            return _offsetRight
        }
        set {
            _offsetRight = newValue
            reloadPointsLayerout()
        }
    }
    
    public var pointWidth: Float {
        get {
            return _pointWidth
        }
        set {
            _pointWidth = newValue
            reloadPointsSize()
        }
    }
    
    public var normalColor: UIColor {
        get {
            return _normalColor
        }
        set {
            _normalColor = newValue
            _normalImg = nil
            reloadPointsStyle()
        }
    }
    public var selectedColor: UIColor {
        get {
            return _selectedColor
        }
        set {
            _selectedColor = newValue
            _selectedImg = nil
            reloadPointsStyle()
        }
    }
    
    public var normalImg: UIImage? {
        get {
            return _normalImg
        }
        set {
            _normalImg = newValue
            _normalColor = UIColor.clear
            reloadPointsStyle()
        }
    }
    public var selectedImg: UIImage? {
        get {
            return _selectedImg
        }
        set {
            _selectedImg = newValue
            _selectedColor = UIColor.clear
            reloadPointsStyle()
        }
    }
    
    public var pageCount: UInt {
        get {
            return _pageCount
        }
        set {
            _pageCount = newValue
            reloadControl()
        }
    }
    public var currentPage: UInt {
        get {
            return _currentPage
        }
        set {
            _currentPage = newValue
            reloadPointsStyle()
        }
    }
}
